counterfactuals.qmd

```diff
@@ -0,0 +1,276 @@
+
+```{r}
+sys <- dust_system_create(sis, pars, n_particles = 1, n_groups = length(pars),
+                          time = time_start, dt = dt)
+
+dust_system_set_state_initial(sys)
+t <- seq(0, 150)
+res <- dust_system_simulate(sys, t)
+traj_counterfactual <- dust_unpack_state(sys, res)
+```
+
+We can then plot the results to compare the fitted trajectories to the counterfactual ones.
+
+```{r}
+incidence_fit <- array(traj_fit$incidence, c(150, 200))
+matplot(data$time, incidence_fit, type = "l", col = "#00000044", lty = 1,
+        xlab = "Time", ylab = "Incidence")
+matlines(t, t(traj_counterfactual$incidence), col = "blue")
+points(data, pch = 19, col = "red")
+```
+
+
+### Running counterfactuals from a timepoint in the data time series
+
+Suppose now we want to examine what would have happened if schools had not opened at `time = 60` (as they had in our fitting) and had instead stayed closed until `time = 130`. This counterfactual would diverge from the fitted trajectories at `time = 60`. To implement this counterfactual we need to save a "snapshot" of the state at `time = 60`.
+
+```{r}
+likelihood <- dust_likelihood_monty(filter, packer, save_trajectories = TRUE,
+                                    save_snapshots = 60)
+
+posterior <- likelihood + prior
+
+samples <- monty_sample(posterior, sampler, 500, initial = c(0.3, 0.1, 0.5),
+                        n_chains = 4)
+samples <- monty_samples_thin(samples, burnin = 100, thinning_factor = 8)
+traj_fit <- dust_unpack_state(filter, samples$observations$trajectories)
+
+```
+
+We will unpack the fitted parameters into a list (of length equal to the sample size) of lists of parameters
+
+```{r}
+pars <- array(samples$pars, c(3, 200))
+pars <- lapply(seq_len(200), function(i) packer$unpack(pars[, i]))
+```
+
+and use a function to change `schools_time` and `schools_open` within each list of parameters
+
+```{r}
+f <- function(p) {
+  p$schools_time <- c(0, 50, 130, 170, 180)
+  p$schools_open <- c(1, 0, 1, 0, 1)
+  p
+}
+pars <- lapply(pars, f)
+```
+
+We create a dust system setting the time to the start time of our counterfactual, set the state of the system using our snapshot (reshaping to remove the chains dimension first) and simulate.
+
+```{r}
+sys <- dust_system_create(sis, pars, n_particles = 1, n_groups = length(pars),
+                          time = 60, dt = dt)
+
+snapshot <- array(samples$observations$snapshots, c(3, 200))
+dust_system_set_state(sys, snapshot)
+
+t <- seq(60, 150)
+res <- dust_system_simulate(sys, t)
+traj_counterfactual <- dust_unpack_state(sys, res)
+```
+
+We can then plot the results to compare the fitted trajectories to the counterfactual ones.
+
+```{r}
+incidence_fit <- array(traj_fit$incidence, c(150, 200))
+matplot(data$time, incidence_fit, type = "l", col = "#00000044", lty = 1,
+        xlab = "Time", ylab = "Incidence")
+matlines(t, t(traj_counterfactual$incidence), col = "blue")
+points(data, pch = 19, col = "red")
+```
```

Comment:
Perhaps add a bit more here on details, specifically:

* you cannot compute trajectories or state after having run an mcmc
* the trajectories are different to running the model with a set of parameters